### PR TITLE
Fix Rename issue for Cyclic unions

### LINKED ISSUE #29742: Renaming doesn't work as expected with cyclic unions
**Description:**
$title.

**Steps to reproduce:**
Consider the following sample.

```ballerina
type CyclicUnion int|CyclicUnion[];

public function main() {
    CyclicUnion x = 1;
}
```

Renaming `CyclicUnion` to `Union` in the variable declaration renames as follows.
```ballerina
type CyclicUnion int|Union[];

public function main() {
    Union x = 1;
}
```

Changing the name in the type definition doesn't change anything else.
```ballerina
type Union int|CyclicUnion[];

public function main() {
    CyclicUnion x = 1;
}
```

Renaming the member of the union doesn't do anything.

**Affected Versions:**
slalpha4-SNAPSHOT

### PR DESCRIPTION
**Pull request by KRVPerera**
Fixes #29742

## Check List 
- [x] Read the [Contributing Guide](https://github.com/ballerina-platform/ballerina-lang/blob/master/CONTRIBUTING.md)
- [ ] Updated Change Log
- [ ] Checked Tooling Support (#<Issue Number>)
- [ ] Added necessary tests
  - [x] Unit Tests
  - [ ] Spec Conformance Tests
  - [ ] Integration Tests
  - [ ] Ballerina By Example Tests
- [ ] Increased Test Coverage   
- [ ] Added necessary documentation  
  - [ ] API documentation 
  - [ ] Module documentation in Module.md files
  - [ ] Ballerina By Examples

### CODE REVIEW

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/SymbolEnter.java [hunk 1]
```diff
@@ -1451,8 +1451,10 @@ public void visit(BLangTypeDefinition typeDefinition) {
         definedType.flags |= typeDefSymbol.flags;
 
         typeDefinition.symbol = typeDefSymbol;
-
-        if (!typeDefinition.hasCyclicReference) {
+        if (typeDefinition.hasCyclicReference) {
+            // Workaround for https://github.com/ballerina-platform/ballerina-lang/issues/29742
+            typeDefinition.type.tsymbol = typeDefSymbol;
```
Comment (KRVPerera) [inline compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/SymbolEnter.java:1456]: Strange but calling `createLabelSymbol` we have created a copy of the type symbol. I am setting the copied symbol back to the type.

**Comment by nadeeshaan** [comment]
Please add a test case for SemanticAPI as well

**Comment by hasithaa** [comment]
@pubudu91 @nadeeshaan @IMS94 Can you check this PR

**Review by nadeeshaan — APPROVED** [review]
Tested locally and the changes looks good.

**Review by hasithaa — APPROVED** [review]

**Review by pubudu91 — COMMENTED** [review]
LGTM. There are a couple of doc issues in the test copyright and description though.

**Comment by KRVPerera** [comment]
> LGTM. There are a couple of doc issues in the test copyright and description though.

ack. Will fix in a subsequent pr

**Pull request merged**